### PR TITLE
write dequantization scripts for DeepSeek V4 FP4/FP8 weights

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

Adds scripts for dequantizing the DeepSeek V4 weights to bf16. DeepSeek gives us: MoE weights in INT8 (actually two FP4s stored in 1 byte because torch doesn't have an FP4 dtype) and attention weights in F8_E4M3. All scaling factors are in F8_E8M0. We heavily reference the huggingface dequantization script [here](https://github.com/huggingface/transformers/blob/da6c53e431f7c9ef0691239d4ce89b0f711ecad7/src/transformers/integrations/finegrained_fp8.py#L933-L1046).

FIXES: b/510020740

# Tests

The [test script](https://paste.googleplex.com/5917486852734976) runs inference on deepseek v4 flash through the transformers library, loading weights from the original deepseek checkpoints and our script dequantized checkpoints, then comparing kl-div, token output, etc. Test results can be seen [here](https://paste.googleplex.com/5920147282984960).

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+    print(f"Successfully saved dequantized BF16 model to: {output_path}")
+
+
+if __name__ == "__main__":
+    parser = ArgumentParser(description="Dequantize DeepSeek hybrid checkpoints (FP8/FP4) to BF16.")
+    parser.add_argument("--input-path", "--input-fp8-hf-path", type=str, required=True,
+                        help="Path to DeepSeek FP8/FP4 Hugging Face folder")
+    parser.add_argument("--output-path", "--output-bf16-hf-path", type=str, required=True,
+                        help="Directory to save output BF16 weights")
```
Comment (parambole) [inline src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py]: nit: We can probably have a check to make sure that the input and output paths are not the same ( to prevent override ).

File: src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+from safetensors.torch import load_file, save_file
+
+# Lookup table for E2M1 FP4 (Two e2m1 nibbles packed per int8/uint8 byte)
+_FP4_E2M1_LUT = torch.tensor(
+    [0.0, 0.5, 1.0, 1.5, 2.0, 3.0, 4.0, 6.0, -0.0, -0.5, -1.0, -1.5, -2.0, -3.0, -4.0, -6.0],
+    dtype=torch.float32
+)
+
+
+def weight_dequant_cpu(x: torch.Tensor, s: torch.Tensor, block_size: int = 128) -> torch.Tensor:
```
Comment (parambole) [inline src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py]: Rename it to dequantize_fp8 ?

File: src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    """
+    assert x.dim() == 2 and s.dim() == 2, "Both x and s must be 2D tensors"
+
+    M, N = x.shape
+
+    x = x.to(torch.float32)
+    y = torch.empty_like(x, dtype=torch.bfloat16)
+
+    for i in range(0, M, block_size):
+        for j in range(0, N, block_size):
```
Comment (parambole) [inline src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py]: nit: Can we vectorize this operation ?

**Review by parambole — COMMENTED** [review]
Thank you for adding this script. I have left a few comments PTAL.

**Review by entrpn — APPROVED** [review]

**Review by parambole — COMMENTED** [review]
`File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md`
`File: tests/end_to_end/tpu/kimi/Run_Kimi.md`


Broken references to deleted script. This documentation file still references the deleted deepseek_fp8_to_bf16.py script. It should be updated to refer to deepseek_dequantize.py and use the new argument names (--input-path and --output-path).

**Review by parambole — APPROVED** [review]
LGTM

File: src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
+r"""Convert weights from FP8/FP4 to BF16 for a DeepSeek HF model.
```
Comment (shuningjin) [inline src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py:15]: nit: Missing Apache License Copyright Header

File: src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    return (x * s_expanded.to(torch.float32)).to(torch.bfloat16)
+
+  # Reshape and broadcast
+  q = x.reshape(M // block_size, block_size, N // block_size, block_size)
+  scale = s.to(torch.float32).reshape(M // block_size, 1, N // block_size, 1)
+
+  return (q * scale).to(torch.bfloat16).reshape(M, N)
+
+
+# reference: huggingface/transformers/blob/main/src/transformers/integrations/finegrained_fp8.py
```
Comment (shuningjin) [inline src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py]: nit: permalink
```
reference: https://github.com/huggingface/transformers/blob/9a0fe3f5dd36ffe1888133f09eb03f1eb14b8a6e/src/transformers/integrations/finegrained_fp8.py#L987
```

File: src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+  """
+  u8 = packed.contiguous().view(torch.uint8)
+  low = (u8 & 0xF).long()
+  high = ((u8 >> 4) & 0xF).long()
+
+  unpacked = torch.stack([_FP4_E2M1_LUT[low], _FP4_E2M1_LUT[high]], dim=-1)
+  return unpacked.reshape(*packed.shape[:-1], 2 * packed.shape[-1])
+
+
+# reference: huggingface/transformers/blob/main/src/transformers/integrations/finegrained_fp8.py
```
Comment (shuningjin) [inline src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py]: nit: permalink
```
reference: https://github.com/huggingface/transformers/blob/9a0fe3f5dd36ffe1888133f09eb03f1eb14b8a6e/src/transformers/integrations/finegrained_fp8.py#L996
```

File: src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  parser = ArgumentParser(description="Dequantize DeepSeek hybrid checkpoints (FP8/FP4) to BF16.")
+  parser.add_argument(
+      "--input-path", "--input-fp8-hf-path", type=str, required=True, help="Path to DeepSeek FP8/FP4 Hugging Face folder"
+  )
+  parser.add_argument(
+      "--output-path", "--output-bf16-hf-path", type=str, required=True, help="Directory to save output BF16 weights"
+  )
+  parser.add_argument(
+      "--cache-size", "--cache-file-num", type=int, default=2, help="Max cached files in RAM during indexing lookup"
+  )
```
Comment (shuningjin) [inline src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py]: nit: Let's use one name for simplicity: "--input-path", "--output-path", "--cache-size"

File: src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
+r"""Convert weights from FP8/FP4 to BF16 for a DeepSeek HF model.
+
```
Comment (shuningjin) [inline src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py:16]: Could you further clarify the usage in docstring?
```
The script can convert quantized weight in the following models:
- DeepSeek-V3, DeepSeek-V3.2 (fp8)
- Kimi-K2-Instruct (fp8)
- DeepSeek-V4-Flash-Base, DeepSeek-V4-Pro-Base (fp8)
- DeepSeek-V4-Flash, DeepSeek-V4-Pro (fp4 + fp8)
```

File: src/maxtext/checkpoint_conversion/utils/param_mapping.py
Comment (shuningjin) [inline src/maxtext/checkpoint_conversion/utils/param_mapping.py]: All changes in `param_mapping.py` seem irrelevant formatting. Could revert.

File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md [hunk 9]
```diff
@@ -44,10 +44,10 @@ hf download deepseek-ai/DeepSeek-V3.2 --local-dir <local_fp8_path>
 ```
 
 #### 2. Dequantize Weights
-Convert the weights from FP8 to BF16 using script [deepseek_fp8_to_bf16.py](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_fp8_to_bf16.py) on CPU:  
+Convert the weights from FP8 to BF16 using script [deepseek_dequantize.py](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py) on CPU:  
 
 ```bash
-python3 -m maxtext.checkpoint_conversion.standalone_scripts.deepseek_fp8_to_bf16 --input-fp8-hf-path=<local_fp8_path> --output-bf16-hf-path=<local_bf16_path>  
+python3 -m maxtext.checkpoint_conversion.standalone_scripts.deepseek_dequantize --input-path=<local_fp8_path> --output-path=<local_bf16_path>
```
Comment (shuningjin) [inline tests/end_to_end/tpu/deepseek/Run_DeepSeek.md:50]: Thanks for updating dequantization command in deepseek user guide! Could you also update kimi:
https://github.com/AI-Hypercomputer/maxtext/blob/dfd8d293d266fe224b90f7cb0b49f3e8084e9892/tests/end_to_end/tpu/kimi/Run_Kimi.md?plain=1#L32-L35

File: src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+  if rows % scale_rows != 0 or cols % scale_cols != 0:
+    raise ValueError(f"Weight shape ({rows}, {cols}) not divisible by scale grid ({scale_rows}, {scale_cols}).")
+
+  block_m = rows // scale_rows
+  block_n = cols // scale_cols
+
+  original_shape = quantized_fp32.shape
+  q = quantized_fp32.reshape(-1, scale_rows, block_m, scale_cols, block_n)
+  s = scales.to(torch.float32).reshape(-1, scale_rows, scale_cols).unsqueeze(-1).unsqueeze(2)
```
Comment (shuningjin) [inline src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py]: Do we need `uint8` path?
https://github.com/huggingface/transformers/blob/9a0fe3f5dd36ffe1888133f09eb03f1eb14b8a6e/src/transformers/integrations/finegrained_fp8.py#L1029-L1035

**Review by shuningjin — COMMENTED** [review]
Thank you! Overall looks good. Some minor comments.

**Review by shuningjin — APPROVED** [review]
Thank you for upgrading the de-quantization support!

**Review by aireenmei — APPROVED** [review]
